src/serprog.c

```diff
@@ -0,0 +1,510 @@
+/*
+ * avrdude - A Downloader/Uploader for AVR device programmers
+ * Support for using serprog programmers to program over ISP.
+ * For information on serprog see:
+ * https://flashrom.org/supported_hw/supported_prog/serprog/index.html
+ *
+ * Copyright (C) 2024 Sydney Louisa Wilke <git@funkeleinhorn.com>
+ * used linuxspi.c as a template:
+ *    Copyright (C) 2013 Kevin Cuzner <kevin@kevincuzner.com>
+ *    Copyright (C) 2018 Ralf Ramsauer <ralf@vmexit.de>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ * known limitations:
+ *  - performance is suboptimal
```

Comment:
Don't forget to remove this line once you've implemented paged r/w. I don't think you can do any better than paged r/w, as the protocol requires at least you send the bytes (with the n-bytes r/w of serprog) and comms will be the limiting factor, timewise. Is `serprog` able to make use of high serial baud rates? That would be another way of sppeding things up, as AVRDUDE, at least under linux, can send/receive at high spped (`-B`)

Reply:
The supported serial baud rate is depending on the serprog programmer used. pico-serprog should be able to do USB full speed so that the serial baud rate shouldn't be the limiting factor. I looked into implementations of paged read and write and how they work at a protocol level and I think I figured out how to implement them. I will try to do this when I find the time for it as I am quite busy at the moment. If you want I can also do this in a follow up PR so the programmer can already be used and tested even though it is not optimal yet.